### PR TITLE
Problem: websocket rpc method id may overlap

### PR DESCRIPTION
**Pull request by yihuang**
Solution: Use sequencial atomic variable to generate unique method ids.

### CODE REVIEW

**Comment by yihuang** [comment]
> I'll suggest to get rid of `HashMap` entirely and use [slab](https://crates.io/crates/slab) instead. `Slab` is the default choice for many networking crates. Any comments?

Yeah, slab makes sense here.

**Comment by yihuang** [comment]
> > I'll suggest to get rid of `HashMap` entirely and use [slab](https://crates.io/crates/slab) instead. `Slab` is the default choice for many networking crates. Any comments?
> 
> Yeah, slab makes sense here.

`slab` reusing method ids is a problem, when receiver timeout, it removed the method id from slab, the same id gets reused by another request, then two responses with the same id may arrive.

File: client-common/src/tendermint/websocket_rpc_client.rs [hunk 1]
```diff
@@ -126,7 +125,8 @@ impl WebsocketRpcClient {
         method: &str,
         params: &[Value],
     ) -> Result<(String, Receiver<JsonRpcResponse>)> {
-        let (message, id) = prepare_message(method, params)?;
+        let id = self.unique_id.fetch_add(1, Ordering::Relaxed).to_string();
```
Comment (devashishdxt) [inline client-common/src/tendermint/websocket_rpc_client.rs]:
```suggestion
        let id = self.unique_id.fetch_add(1, Ordering:: SeqCst).to_string();
```

Comment (yihuang) [reply client-common/src/tendermint/websocket_rpc_client.rs]: As I understands, `Relaxed` is enough to keep it unique.

Comment (devashishdxt) [reply client-common/src/tendermint/websocket_rpc_client.rs]: `Relaxed` performs almost no synchronisation. In our case, this value can be accessed from multiple threads, so, it'd be good to use `Ordering::SeqCst`. You can read more about this [here](https://en.cppreference.com/w/cpp/atomic/memory_order).

Comment (yihuang) [reply client-common/src/tendermint/websocket_rpc_client.rs]: > `Relaxed` performs almost no synchronisation. In our case, this value can be accessed from multiple threads, so, it'd be good to use `Ordering::SeqCst`. You can read more about this [here](https://en.cppreference.com/w/cpp/atomic/memory_order).

We only need the atomicity of the increment operation to get the unique id, no needs to synchronize with other operations, so I think it's fine with `Relaxed` here.

Comment (yihuang) [reply client-common/src/tendermint/websocket_rpc_client.rs]: I find some explanations [here](https://vorner.github.io/2018/03/25/Atomics.html#relaxed-order--sanity-just-for-me).

Comment (devashishdxt) [reply client-common/src/tendermint/websocket_rpc_client.rs]: If there are two separate requests from different threads, can you explain how will this work?

Comment (yihuang) [reply client-common/src/tendermint/websocket_rpc_client.rs]: > If there are two separate requests from different threads, can you explain how will this work?

No matter what the `Ordering` is, the operation is atomic, what's what `Atomic` always guarantee.
So two concurrent request happens, either one gets 1 and the other gets 2, or one gets 2 and the other gets 1.

Comment (devashishdxt) [reply client-common/src/tendermint/websocket_rpc_client.rs]: Ok

**Pull request merged**